### PR TITLE
METRON-1122: Add support for the profiler in the management pack

### PR DESCRIPTION
**Pull request by cestella**
## Contributor Comments
Name says it all: Add support for the profiler in the management pack

## Pull Request Checklist

Thank you for submitting a contribution to Apache Metron.  
Please refer to our [Development Guidelines](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=61332235) for the complete guide to follow for contributions.  
Please refer also to our [Build Verification Guidelines](https://cwiki.apache.org/confluence/display/METRON/Verifying+Builds?show-miniview) for complete smoke testing guides.  

In order to streamline the review of the contribution we ask you follow these guidelines and ask you to double check the following:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? If not one needs to be created at [Metron Jira](https://issues.apache.org/jira/browse/METRON/?selectedTab=com.atlassian.jira.jira-projects-plugin:summary-panel). 
- [x] Does your PR title start with METRON-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.
- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

### For code changes:
- [x] Have you included steps to reproduce the behavior or problem that is being changed or addressed?
- [x] Have you included steps or a guide to how the change may be verified and tested manually?
- [x] Have you ensured that the full suite of tests and checks have been executed in the root metron folder via:
  ```
  mvn -q clean integration-test install && build_utils/verify_licenses.sh 
  ```

- [x] Have you written or updated unit tests and or integration tests to verify your changes?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [x] Have you verified the basic functionality of the build by building and running locally with Vagrant full-dev environment or the equivalent?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered by building and verifying the site-book? If not then run the following commands and the verify changes via `site-book/target/site/index.html`:

  ```
  cd site-book
  mvn site
  ```

#### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.
It is also recommended that [travis-ci](https://travis-ci.org) is set up for your personal repository such that your branches are built there before submitting a pull request.

### CODE REVIEW

File: metron-deployment/packaging/ambari/metron-mpack/src/main/resources/addon-services/METRON/CURRENT/role_command_order.json [hunk 1]
```diff
@@ -5,11 +5,13 @@
         "_comment" : "dependencies for all cases",
         "METRON_INDEXING-INSTALL" : ["METRON_PARSERS-INSTALL"],
         "METRON_ENRICHMENT-INSTALL": ["METRON_INDEXING-INSTALL"],
+        "METRON_PROFILER-INSTALL": ["METRON_ENRICHMENT-INSTALL"],
         "METRON_REST-INSTALL": ["METRON_PARSERS-INSTALL"],
         "METRON_PARSERS-START" : ["NAMENODE-START", "ZOOKEEPER_SERVER-START", "KAFKA_BROKER-START", "STORM_REST_API-START","METRON_ENRICHMENT_MASTER-START"],
         "METRON_ENRICHMENT_MASTER-START" : ["NAMENODE-START", "ZOOKEEPER_SERVER-START", "KAFKA_BROKER-START", "STORM_REST_API-START", "HBASE_MASTER-START", "HBASE_REGIONSERVER-START"],
         "METRON_ENRICHMENT_SERVICE_CHECK-SERVICE_CHECK" : ["METRON_ENRICHMENT_MASTER-START"],
         "METRON_INDEXING-START" : ["NAMENODE-START", "ZOOKEEPER_SERVER-START", "KAFKA_BROKER-START", "STORM_REST_API-START","METRON_PARSERS-START"],
+        "METRON_PROFILER-START" : ["NAMENODE-START", "ZOOKEEPER_SERVER-START", "KAFKA_BROKER-START", "STORM_REST_API-START","METRON_INDEXING-START"],
```
Comment (simonellistonball) [inline metron-deployment/packaging/ambari/metron-mpack/src/main/resources/addon-services/METRON/CURRENT/role_command_order.json]: does it really need indexing running to start? (to work, yes, but to start?)

File: metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/configuration/metron-profiler-env.xml [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+-->
+<configuration supports_final="true">
+  <property>
+    <name>profiler_kafka_start</name>
```
Comment (simonellistonball) [inline metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/configuration/metron-profiler-env.xml]: Should this be a select-option choice like the enrichments section?  (https://github.com/apache/metron/blob/ae1dc3a4aa31cfd6b4af7558dbb2fc0314f80a29/metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/configuration/metron-enrichment-env.xml#L36)

File: metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/configuration/metron-profiler-env.xml [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    <description>One of EARLIEST, LATEST, UNCOMMITTED_EARLIEST, UNCOMMITTED_LATEST</description>
+    <display-name>Input Topic Start</display-name>
+  </property>
+  <property>
+    <name>profiler_period_duration</name>
+    <value>15</value>
+    <description>The duration of each profile period. This value should be defined along with profiler.period.duration.units</description>
+    <display-name>Period Duration</display-name>
+  </property>
+  <property>
```
Comment (simonellistonball) [inline metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/configuration/metron-profiler-env.xml]: Should constrain choice here too

File: metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/configuration/metron-profiler-env.xml [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    <name>profiler_ttl</name>
+    <value>30</value>
+    <description>If a message has not been applied to a Profile in this period of time, the Profile will be terminated and its resources will be cleaned up. This value should be defined along with profiler.ttl.units.
+      This time-to-live does not affect the persisted Profile data in HBase. It only affects the state stored in memory during the execution of the latest profile period. This state will be deleted if the time-to-live is exceeded.
+    </description>
+    <display-name>Time to Live</display-name>
+  </property>
+  <property>
+    <name>profiler_ttl_units</name>
+    <value>MINUTES</value>
```
Comment (simonellistonball) [inline metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/configuration/metron-profiler-env.xml]: Should constrain choice here too

File: metron-deployment/packaging/ambari/metron-mpack/src/main/resources/addon-services/METRON/CURRENT/role_command_order.json (hunk 1 above)
Comment (cestella) [inline metron-deployment/packaging/ambari/metron-mpack/src/main/resources/addon-services/METRON/CURRENT/role_command_order.json]: Yeah, you're right.  It's actually worse than what you suggest, it's overtly wrong: it needs zookeeper, storm, kafka, and hbase

File: metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/configuration/metron-profiler-env.xml (hunk 2 above)
Comment (cestella) [inline metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/configuration/metron-profiler-env.xml]: +1, good catch.  I'm a dork.

File: metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/configuration/metron-profiler-env.xml (hunk 3 above)
Comment (cestella) [inline metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/configuration/metron-profiler-env.xml]: +1, good catch. I'm a dork.

File: metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/configuration/metron-profiler-env.xml (hunk 4 above)
Comment (cestella) [inline metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/configuration/metron-profiler-env.xml]: +1, good catch. I'm a dork.

File: metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/package/scripts/params/status_params.py [hunk 5]
```diff
@@ -52,6 +52,26 @@
 update_table = 'metron_update'
 update_cf = 't'
 
+# Profiler
+metron_profiler_topology = 'profiler'
+profiler_input_topic = config['configurations']['metron-enrichment-env']['enrichment_output_topic']
+profiler_kafka_start = config['configurations']['metron-profiler-env']['profiler_kafka_start']
+profiler_period_duration = config['configurations']['metron-profiler-env']['profiler_period_duration']
+profiler_period_units = config['configurations']['metron-profiler-env']['profiler_period_units']
+profiler_ttl = config['configurations']['metron-profiler-env']['profiler_ttl']
```
Comment (merrimanr) [inline metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/package/scripts/params/status_params.py]: Do all of these need to be defined here?  These are also defined in params_linux.py and most won't be used for status.

File: metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/package/scripts/params/params_linux.py [hunk 6]
```diff
@@ -202,6 +203,7 @@
 client_jaas_path = metron_home + '/client_jaas.conf'
 client_jaas_arg = '-Djava.security.auth.login.config=' + metron_home + '/client_jaas.conf'
 enrichment_topology_worker_childopts = client_jaas_arg if security_enabled else ''
+profiler_topology_worker_childopts = client_jaas_arg if security_enabled else ''
```
Comment (merrimanr) [inline metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/package/scripts/params/params_linux.py]: As it stand, setting this property in Ambari won't have any effect.  There needs to be an additional step that reads from the config and appends.  See enrichment_topology_worker_childopts as an example.

File: metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/package/scripts/params/params_linux.py [hunk 7]
```diff
@@ -251,6 +253,26 @@
 threat_intel_join_parallelism = config['configurations']['metron-enrichment-env']['threat_intel_join_parallelism']
 kafka_writer_parallelism = config['configurations']['metron-enrichment-env']['kafka_writer_parallelism']
 
+# Profiler
+
+metron_profiler_topology = 'profiler'
+profiler_input_topic = config['configurations']['metron-enrichment-env']['enrichment_output_topic']
```
Comment (merrimanr) [inline metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/package/scripts/params/params_linux.py]: Was this intentional?  Wouldn't we want a configurable profiler input topic?

File: metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/configuration/metron-profiler-env.xml [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+    </value-attributes>
+  </property>
+  <property>
+    <name>profiler_topology_workers</name>
+    <value>1</value>
+    <description>The profiler storm topology workers</description>
+    <display-name>Number of Workers</display-name>
+  </property>
+  <property>
+    <name>profiler_topology_executors</name>
```
Comment (merrimanr) [inline metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/configuration/metron-profiler-env.xml]: This property is confusing to me.  It ends up being bound to the Storm topology.acker.executors property in the profiler flux file.  The indexing and enrichment topologies also have this property but it's named differently (indexing_acker_executors in the mpack and Enrichment Ackers in the Ambari UI for example).  Whatever the correct naming convention is, I think they should be consistent.

File: metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/themes/metron_theme.json [hunk 9]
```diff
@@ -214,6 +214,73 @@
             }
           },
           {
+            "name": "profiler",
```
Comment (merrimanr) [inline metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/themes/metron_theme.json]: I noticed the layout is slightly different in the Ambari UI than other tabs (Enrichment for example).  I don't think it necessarily looks bad (might even be an improvement) but ideally all the tabs are consistent.

File: metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/package/scripts/params/status_params.py (hunk 5 above)
Comment (ottobackwards) [reply metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/package/scripts/params/status_params.py]: Just a side question @merrimanr, how do you know that?  Do you have documentation on what the status scripts are, when and who loads them?  A guide?  Asking for a friend.

Comment (merrimanr) [reply metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/package/scripts/params/status_params.py]: I found out the hard way, through experience.  The ambari agent runs in a polling loop and calls the "status" function on the command script (profiler_master.py in this case).  Not much documentation out there for this as far as I know.

File: metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/package/scripts/params/params_linux.py (hunk 7 above)
Comment (cestella) [reply metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/package/scripts/params/params_linux.py]: So, the profiler's input topic, in metron, is the output of the enrichment topology.  I did it this way to keep them in alignment.  There is currently, to my mind, not a usecase where the profiler does take its data from the output of the enrichment topology.

File: metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/configuration/metron-profiler-env.xml (hunk 8 above)
Comment (cestella) [reply metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/configuration/metron-profiler-env.xml]: Yeah, we named this property differently in the profiler.properties file than in indexing and enrichment and I was following the convention in the properties file.  I can adjust to make it consistent.

File: metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/package/scripts/params/params_linux.py (hunk 6 above)
Comment (cestella) [reply metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/package/scripts/params/params_linux.py]: whoops, missed a step.  Thanks!

File: metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/package/scripts/params/params_linux.py (hunk 7 above)
Comment (nickwallen) [reply metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/package/scripts/params/params_linux.py]: Its a good thing to bring up and discuss @merrimanr .  I agree with @cestella .  Doing it this way makes it less error-prone for the user.  If the user changes the enrichment output topic, the Profiler will get auto-magically updated too.  

I cannot think of a use case where the Profiler does not want the output of Enrichment.

Comment (merrimanr) [reply metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/package/scripts/params/params_linux.py]: That makes sense.  I'm good with it.

File: metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/themes/metron_theme.json (hunk 9 above)
Comment (cestella) [reply metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/themes/metron_theme.json]: Yeah, I agree, we should have a consistent look and feel here.  I'll adjust.  I think the REST panel looks like this and I cut and pasted.  I'll look into it.

File: metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/package/scripts/params/params_linux.py (hunk 7 above)
Comment (cestella) [reply metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/package/scripts/params/params_linux.py]: It probably bears thinking about if we want to do the indexing topology this way too, honestly.

File: metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/package/scripts/params/status_params.py (hunk 5 above)
Comment (cestella) [reply metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/package/scripts/params/status_params.py]: Yeah, I didn't quite entirely realize what the status.py was doing, so I went a bit overboard.  I'll trim them back to just the stuff needed for status checks.

**Comment by nickwallen** [comment]
I have only noticed one small problem.  If I change the enrichment topic, Ambari correctly indicates that the Enrichment Topology needs restarted.  But since the Enrichment topic also impacts the Profiler, the Profiler needs to be restarted.  There is no indication in Ambari that the Profiler needs restarted.

I don't think we even need to address this as part of this PR.  As long as you validate what I am seeing and we log a JIRA, then I am good with this PR being merged.

+1

File: metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/metainfo.xml [hunk 10]
```diff
@@ -119,6 +119,45 @@
...
+              </auto-deploy>
+            </dependency>
+          </dependencies>
+          <commandScript>
+            <script>scripts/profiler_master.py</script>
+            <scriptType>PYTHON</scriptType>
+            <timeout>600</timeout>
+          </commandScript>
+          <configuration-dependencies>
+            <config-type>metron-profiler-env</config-type>
```
Comment (merrimanr) [inline metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/metainfo.xml]: The comment from @nickwallen can be addressed by adding this line here:
`<config-type>metron-enrichment-env</config-type>`

**Comment by merrimanr** [comment]
I tested this in full dev again and it looks good after the recent changes.  @nickwallen brings up a good point and I would prefer we fix it now because it's an easy one line change.  I added a comment with my suggestion and have verified it in full dev.  Pending that addition, I'm +1.  Nice work @cestella.

File: metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/metainfo.xml (hunk 10 above)
Comment (cestella) [reply metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/metainfo.xml]: Thanks!  I'll add it now.

**Comment by cestella** [comment]
Alright, I think I've fixed the enrichment dependency issue.  We should probably spin up in full-dev again to verify.

**Comment by merrimanr** [comment]
I ran this up in full dev and it worked.  However when I was looking at the Profiler tab in Ambari I noticed "Number of Acker Executors" was set to 0.  Shouldn't this be set to 1, similar to enrichment and indexing?

**Comment by cestella** [comment]
@merrimanr yeah, maybe, this was the default before for the profiler, so it stuck.  Do we want to adjust this to be consistent with enrichment or indexing?  I'd vote yes.

**Comment by cestella** [comment]
@merrimanr @nickwallen Ok, I changed the default acker executors to `1`.  I'll spin up again in full-dev, let me know if there's anything else I missed.

**Comment by nickwallen** [comment]
Assuming your test in Full Dev goes smoothly, my +1 stands.  It works great.

**Comment by merrimanr** [comment]
+1

**Pull request closed without merging**